SEP-0012.md

```diff
@@ -0,0 +1,495 @@
+            `.NDCube.wcs.world_axis_physical_types` or
+            `.NDCube.wcs.world_axis_names` of axes for which real world
+            coordinates are desired. ``axes=None`` implies all axes will be
+            returned.
+
+        edges
+            The edges argument helps in returning ``pixel_edges``
+            instead of ``pixel_values``. Default value is False,
+            which returns ``pixel_values``. True return ``pixel_edges``
+
+        wcs
+            The WCS object to used to calculate the world coordinates.
+            Although technically this can be any valid WCS, it will typically be
+            ``self.wcs``, ``self.extra_coords``, or ``self.combined_wcs``, combing both
+            the WCS and extra coords.
+            Default=self.wcs
+
+        Returns
+        -------
+        axes_coords
+            High level object, or iterable thereof, giving the real world
+            coords for the axes requested by user.
+            For example, SkyCoords. The types returned are determined by the WCS object.
+
+        Example
+        -------
+        >>> NDCube.all_world_coords(('lat', 'lon'))
+        >>> NDCube.all_world_coords(2)
+
+        """
+
+    @abc.abstractmethod
+    def axis_world_coords_values(self,
+                                 *axes: Union[int, str],
+                                 edges: bool = False,
+                                 wcs: Optional[Union[BaseHighLevelWCS, ExtraCoordsABC]] = None
+                                 ) -> Iterable[u.Quantity]:
+        """
+        Returns world coordinate values of all pixels for all axes.
+
+        Parameters
+        ----------
+        axes
+            Axis number in numpy ordering or unique substring of
+            `.NDCube.wcs.world_axis_physical_types` or
+            `.NDCube.wcs.world_axis_names` of axes for which real world
+            coordinates are desired. ``axes=None`` implies all axes will be
+            returned.
+
+        edges
+            The edges argument helps in returning ``pixel_edges``
+            instead of ``pixel_values``. Default value is False,
+            which returns ``pixel_values``. True return ``pixel_edges``
+
+        wcs
+            The WCS object to used to calculate the world coordinates.
+            Although technically this can be any valid WCS, it will typically be
+            ``self.wcs``, ``self.extra_coords``, or ``self.combined_wcs``, combing both
+            the WCS and extra coords.
+            Default=self.wcs
+
+        Returns
+        -------
+        axes_coords
+            `~astropy.units.Quantity` or iterable thereof for all requested
+            world axes, units determined by the wcs.  # TODO: Should we mandate namedtuple?
+
+        Example
+        -------
+        >>> NDCube.all_world_coords_values(('lat', 'lon'))
+        >>> NDCube.all_world_coords_values(2)
+
+        """
+
+    @abc.abstractmethod
+    def crop(self,
```

Comment:
Having personally had fun with inputs to `submap` in sunpy, this needs a much tighter definition. In particular

- Given a coordinate that lies in a pixel, is that pixel included or not in the cropped result?
- Are the edges of the cropping done using parallel N-1D planes in pixel space, or coordinate space? My guess is pixel space, but this needs to be explicitly stated.

Reply:
This method currently includes pixels that contain the lower and upper corner values.  The justification is that in array slicing all elements strictly less than the upper index are returned.  However, depending on where in the pixel the ``upper_corner`` value falls, there may be some real world values < ``upper_corner`` AND > ``edge of the returned cube``.  The only way to avoid this it to include ``upper_corner`` in the returned cube. 

I think the way to define he behavior is that this method:
* returns the smallest cube (in pixel space) that includes the entire real world region of interest defined by ``lower_corner`` and ``upper_corner``.
* If the region of interest extends beyond the FOV of the cube in a given dimension, everything up to the edge of the cube along that dimension is returned.  Padding array elements are not generated.
* If there is no overlap between the cube and the region of interest, in `IndexError` is raised.

This behavior is analogous to the behavior of numpy slicing except the the upper limit pixel is included:
```python
>>> np.arange(4)[1:3]  # OK
>>> np.arange(4)[1:10] # OK
>>> np.arange(4)[5:10] # Raises an IndexError

Reply:
How is "entire real world region of interest" defined? Is it a cube in coordinate space? And then you choose the smallest cube in pixel space that contains the cube in coordinate space?

If my interpretation is correct, this is *not* how `submap` in sunpy works - there the cropped area is simply a square in pixel space, with corners given by the `bottom_left` and `top_right` coordinates.

Reply:
> How is "entire real world region of interest" defined? Is it a cube in coordinate space? And then you choose the smallest cube in pixel space that contains the cube in coordinate space?

Yes.  That's exactly it.

> If my interpretation is correct, this is _not_ how `submap` in sunpy works - there the cropped area is simply a square in pixel space, with corners given by the `bottom_left` and `top_right` coordinates. Not an issue, but just thought I'd mention it.

Unless you resample the data I don't think it's generally possible to do it another way to how `NDCube.crop` does it.  The exception would be the degenerate case where all the world coordinates are aligned with the array axes, in which case what `NDCube.crop` does is equivalent.  I didn't think `submap` resampled the data but I could be wrong.

Reply:
I there might be a bit of confusion; this is the way `submap` works:

1. Convert `bottom_left_coord` and `top_right_coord` to array indices
2. Return `data[bottom_left_idx[0]:top_right_idx[0], bottom_left_idx[1], top_right_idx[1]]`

No resampling or or interpolating, because one just returns a cube in pixel space. Am I right in thinking that the method proposed here for `crop` is:

1. Find all pixels with
```
(lower_corner[0] <= coordinate[0] <= upper_corner[0]) and
(lower_corner[1] <= coordinate[1] <= upper_corner[1]) and
...
```
This is not a cube in data/pixel space, and actually needs a much tighter definition to take into account the finite size of pixels, and there might be tricky edge cases with highly non-linear data > coordinate transformations that need to be worked out.

2. Find the smallest cube in data/pixel space that contains all of these pixels

Reply:
From the doc I'm unsure if the interval defined by `lower_corner` and `upper_corner` should be interpreted as `lower_corner[0] <= coordinate[0] <= upper_corner[0]` or `lower_corner[0] <= coordinate[0] < upper_corner[0]` similar to @dstansby comments above. But additionally it's not clear to me what value is used for the comparison, I assume it's the pixel centres but that's not explicitly stated and probably should be if it is?

Reply:
The implementation of crop is very similar, if not the same as @dstansby's description above of `submap`.  Edge cases may be handled differently.  To clarify here is the proposed implementation:

1. Convert `lower_corner_coords` and `upper_corner_coords` to array indices, `lower_corner_idx` and `upper_corner_idx`.  The conversion to array indices always round down to the nearest whole index.  If we think of pixels as bins then in the `i`th dimensions, this means `lower_corner_idx[i]` corresponds to the lower edge of the bin/pixel containing `lower_corner_coords[i]` and `upper_corner_idx[i] + 1` corresponds to the upper edge of the bin/pixel containing `upper_corner_coords[i]`
2. Return `data[lower_corner_idx[0] : upper_corner_idx[0] + 1, ...]` (as so on for higher dimensions).  Note that using `upper_corner_idx[i] + 1` means that the upper corner given by the user will always be included in the returned cube.

### Edge Cases
* If `lower_corner_coords[i]` is below the range of the cube, set `lower_corner_idx = 0`.  So in this case an error is not raised.  `crop` returns as much of the region of interest as it can give the cube's limits.
* If `upper_corner_idx[i]` is greater than the axis length return as much of the region of interest as possible given the cube's limits.  Do not raise an error.  (Analogous to the above bullet on the upper side.)
* If `lower_corner_idx[i]` greater than the axis length, then none of the region of interest along the `i`th axis is within the cube.  Raise an `IndexError`.

@dstansby, @samaloney, does this make it clearer?

Reply:
I have added a comment to the docstring clarifying that this works like submap and is therefore subject to the issues laid out in https://github.com/sunpy/sunpy/issues/4896.

I think we can consider fixing that a feature request for the future.